presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+                    .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+            CreateStatisticAggregationsResult result = createStatisticsAggregation(statistics, columnToSymbolMap);
+
+            Map<Symbol, Expression> projections = result.getProjections();
+            Assignments assignments = Assignments.builder()
+                    .putIdentities(source.getOutputSymbols())
+                    .putAll(projections)
+                    .build();
+            ProjectNode projectNode = new ProjectNode(
+                    idAllocator.getNextId(),
+                    source,
+                    assignments);
+
+            StatisticAggregations.Parts aggregations = result.getAggregations().split(symbolAllocator, metadata.getFunctionRegistry());
+
+            StatisticAggregations partialAggregation = aggregations.getPartialAggregation();
+            ImmutableList.Builder<Symbol> writerOutputSymbols = ImmutableList.builder();
+            writerOutputSymbols.addAll(writerOutputs);
+            partialAggregation.getGroupingSets().forEach(writerOutputSymbols::addAll);
+            writerOutputSymbols.addAll(partialAggregation.getAggregations().keySet());
+
+            PlanNode writerNode = new TableWriterNode(
+                    idAllocator.getNextId(),
+                    projectNode,
+                    target,
+                    symbols,
+                    columnNames,
+                    writerOutputSymbols.build(),
+                    partitioningScheme,
+                    Optional.of(partialAggregation));
+
+            TableFinishNode commitNode = new TableFinishNode(
+                    idAllocator.getNextId(),
+                    writerNode,
+                    target,
+                    commitOutputs,
+                    Optional.of(aggregations.getFinalAggregation()),
+                    Optional.of(result.getDescriptor()));
+
+            return new RelationPlan(commitNode, analysis.getRootScope(), commitOutputs);
+        }
 
-        List<Symbol> outputs = ImmutableList.of(symbolAllocator.newSymbol("rows", BIGINT));
         TableFinishNode commitNode = new TableFinishNode(
                 idAllocator.getNextId(),
-                writerNode,
+                new TableWriterNode(
+                        idAllocator.getNextId(),
+                        source,
+                        target,
+                        symbols,
+                        columnNames,
+                        writerOutputs,
+                        partitioningScheme,
+                        Optional.empty()),
                 target,
-                outputs);
+                commitOutputs,
+                Optional.empty(),
+                Optional.empty());
+        return new RelationPlan(commitNode, analysis.getRootScope(), commitOutputs);
+    }
+
+    private CreateStatisticAggregationsResult createStatisticsAggregation(TableStatisticsMetadata statisticsMetadata, Map<String, Symbol> columnToSymbolMap)
```

Comment:
`LogicalPlanner` is already fairly long. See if you can move these methods and result callers into a separate helper class (in a different file)

Reply:
Extracted to `StatisticsAggregationsPlanner`